Add D-FINE Model into Transformers

VladOS95-cyber opened the pull request:
# What does this PR do?

This PR add D-FINE into the Transformers library. There is a new thing in transformers called modular, which adds new models by creating a modeling_modelname.py file. Since D-FINE updates several RT-DETR arch parts while keeping the rest of the model unchanged, it serves as an ideal use case for this modular approach.

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. Link: https://github.com/huggingface/transformers/issues/35283
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

@qubvel @Rocketknight1 @ArthurZucker 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @ylacombe, @eustlb
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

VladOS95-cyber committed:
add base structure

VladOS95-cyber committed:
add base structure

VladOS95-cyber committed:
add base structure

VladOS95-cyber committed:
add decoder

VladOS95-cyber committed:
add decoder

VladOS95-cyber committed:
add decoder

VladOS95-cyber committed:
update decoder, add encoder

VladOS95-cyber committed:
update decoder, add encoder

VladOS95-cyber committed:
update decoder, add encoder

VladOS95-cyber committed:
temp remove

VladOS95-cyber committed:
temp remove

VladOS95-cyber committed:
temp remove

VladOS95-cyber committed:
add encoder

VladOS95-cyber committed:
add encoder

VladOS95-cyber committed:
add encoder

VladOS95-cyber committed:
add docs

VladOS95-cyber committed:
add docs

VladOS95-cyber committed:
add docs

VladOS95-cyber committed:
fix in doc

VladOS95-cyber committed:
fix in doc

VladOS95-cyber committed:
fix in doc

VladOS95-cyber committed:
add generated modeling files, add resnet

VladOS95-cyber committed:
add generated modeling files, add resnet

VladOS95-cyber committed:
add generated modeling files, add resnet

VladOS95-cyber committed:
fix mapping

VladOS95-cyber committed:
fix mapping

VladOS95-cyber committed:
fix mapping

VladOS95-cyber committed:
add init, some fixes

VladOS95-cyber committed:
add init, some fixes

VladOS95-cyber committed:
add init, some fixes

VladOS95-cyber committed:
some fixes

VladOS95-cyber committed:
some fixes

VladOS95-cyber committed:
some fixes

VladOS95-cyber committed:
some fixes for auto and model arch

VladOS95-cyber committed:
some fixes for auto and model arch

VladOS95-cyber committed:
some fixes for auto and model arch

VladOS95-cyber committed:
add test conversion script, some logical fixes, doc update

VladOS95-cyber committed:
add test conversion script, some logical fixes, doc update

VladOS95-cyber committed:
add test conversion script, some logical fixes, doc update

Rocketknight1 commented:
cc @qubvel @NielsRogge

VladOS95-cyber committed:
fix embeddings, fix config, fix DFineForObjectDetection

VladOS95-cyber committed:
fix embeddings, fix config, fix DFineForObjectDetection

VladOS95-cyber committed:
fix embeddings, fix config, fix DFineForObjectDetection

VladOS95-cyber committed:
add config for stages, fix convertion logic and resnet

VladOS95-cyber committed:
add config for stages, fix convertion logic and resnet

VladOS95-cyber committed:
add config for stages, fix convertion logic and resnet

VladOS95-cyber committed:
covertion script fixes, arch logci fixes

VladOS95-cyber committed:
covertion script fixes, arch logci fixes

VladOS95-cyber committed:
covertion script fixes, arch logci fixes

VladOS95-cyber committed:
some fixes for forward prop

VladOS95-cyber committed:
some fixes for forward prop

VladOS95-cyber committed:
some fixes for forward prop

VladOS95-cyber committed:
fix decoder forward prop

VladOS95-cyber committed:
fix decoder forward prop

VladOS95-cyber committed:
fix decoder forward prop

VladOS95-cyber committed:
small refact for convert script

VladOS95-cyber committed:
small refact for convert script

VladOS95-cyber committed:
small refact for convert script

VladOS95-cyber committed:
apply format

VladOS95-cyber committed:
apply format

VladOS95-cyber committed:
apply format

VladOS95-cyber committed:
some fixes for resn net stages forward prop

VladOS95-cyber committed:
some fixes for resn net stages forward prop

VladOS95-cyber committed:
some fixes for resn net stages forward prop

VladOS95-cyber committed:
final fixes to model forward prop

VladOS95-cyber committed:
final fixes to model forward prop

VladOS95-cyber committed:
final fixes to model forward prop

VladOS95-cyber committed:
fix resnet and model modulling

VladOS95-cyber committed:
fix resnet and model modulling

VladOS95-cyber committed:
fix resnet and model modulling

VladOS95-cyber committed:
small fixes in conv layer in decoder

VladOS95-cyber committed:
small fixes in conv layer in decoder

VladOS95-cyber committed:
small fixes in conv layer in decoder

VladOS95-cyber committed:
finalize conversion script, add usage example in doc

VladOS95-cyber committed:
finalize conversion script, add usage example in doc

VladOS95-cyber committed:
finalize conversion script, add usage example in doc

VladOS95-cyber committed:
format fix

VladOS95-cyber committed:
format fix

VladOS95-cyber committed:
format fix

VladOS95-cyber commented:
Model implementation is done. Unit tests are still missing. I am working on it

VladOS95-cyber committed:
couple small fixes

VladOS95-cyber committed:
couple small fixes

VladOS95-cyber committed:
couple small fixes

VladOS95-cyber committed:
fixes for doc, add loss, init tests

VladOS95-cyber committed:
fixes for doc, add loss, init tests

VladOS95-cyber committed:
fixes for doc, add loss, init tests

VladOS95-cyber committed:
fixes decoder to return cross attn

VladOS95-cyber committed:
fixes decoder to return cross attn

VladOS95-cyber committed:
fixes decoder to return cross attn

VladOS95-cyber committed:
add initial model tests

VladOS95-cyber committed:
add initial model tests

VladOS95-cyber committed:
add initial model tests

qubvel reviewed (commented):
Hi @VladOS95-cyber! Thanks a lot for working on the model, super excited to see it merged! It’s a state-of-the-art real-time object detection model at the moment 🚀  

Before diving into the implementation details, here are some general comments we need to address in the PR:  

- [ ] **Integration Tests**  
   Let's set up integration tests to ensure output consistency is maintained while refactoring the modeling code. Please use `torch.testing.assert_close` when comparing tensors (**not use** `self.assertTrue(torch.allclose(...)`)

- [ ] **Modeling Code & RT-DETRv2**  
   Great job keeping using modular! I see it's based on RT-DETR, but RT-DETRv2 was just merged yesterday. It might be helpful, especially since I noticed the `v2` attention is used. 

 - [ ] **Model Organization**  
    <del>Each model should have its own folder. We should separate the backbone model and the D-Fine model, following the same structure as RT-DETR.</del>  Checked RT-DETR, it's actually 2 models in one folder, so we can leave it as is (in case you didn't change yet)

- [ ] **Consistent Naming for Classes & Modules**  
   - Model names and types: `DFine`  and `d_fine`, `DFineResnet` and `d_fine_resnet`.  
   - All module names should be prefixed with the model name. For example, instead of `ResidualBlock`, use `DFineResnetResidualBlock`.  

- [ ] **Code Paths & Cleanup**  
   Remove unused conditional statements if no existing configuration requires them. Also, remove the unnecessary config parameters.

     ```python
     # Instead of:
     if config.use_last_layer_norm:
         self.norm = nn.LayerNorm()
     else:
         self.norm = nn.Identity()
     ```
     ```python
     # If `use_last_layer_norm = True` in all configs, simplify it to:  
     self.norm = nn.LayerNorm()
     ```

- [ ] **Code Style**  
   - Use clear, descriptive variable names: avoid single-letter (`x`) or placeholder (`tmp`) variables.  
   - Add comments for non-obvious code.  
   - Use type hints for modules (e.g. in `__init__` and `forward`).  

- [ ] **Conversion Script Standard**  
   Please follow the `mllama` model format for the conversion script. Define a single key mapping using regex (see `rt_detr_v2`, `ijepa`, `mllama`, `superglue` models for reference). This is our new standard for all newly added models.  

Thanks again! Looking forward to the updates 🚀

VladOS95-cyber committed:
fix conversion logic, add integration tests

VladOS95-cyber committed:
fix conversion logic, add integration tests

VladOS95-cyber committed:
fix conversion logic, add integration tests

VladOS95-cyber committed:
some refactoring for naming and description

VladOS95-cyber committed:
some refactoring for naming and description

VladOS95-cyber committed:
some refactoring for naming and description

VladOS95-cyber commented:
> Hi @VladOS95-cyber! Thanks a lot for working on the model, super excited to see it merged! It’s a state-of-the-art real-time object detection model at the moment 🚀
> 
> Before diving into the implementation details, here are some general comments we need to address in the PR:
> 
> * [ ]  **Integration Tests**
>   Let's set up integration tests to ensure output consistency is maintained while refactoring the modeling code. Please use `torch.testing.assert_close` when comparing tensors (**not use** `self.assertTrue(torch.allclose(...)`)
> * [ ]  **Modeling Code & RT-DETRv2**
>   Great job keeping using modular! I see it's based on RT-DETR, but RT-DETRv2 was just merged yesterday. It might be helpful, especially since I noticed the `v2` attention is used.
> * [ ]  **Model Organization**
>   ~Each model should have its own folder. We should separate the backbone model and the D-Fine model, following the same structure as RT-DETR.~  Checked RT-DETR, it's actually 2 models in one folder, so we can leave it as is (in case you didn't change yet)
> * [ ]  **Consistent Naming for Classes & Modules**
>   
>   * Model names and types: `DFine`  and `d_fine`, `DFineResnet` and `d_fine_resnet`.
>   * All module names should be prefixed with the model name. For example, instead of `ResidualBlock`, use `DFineResnetResidualBlock`.
> * [ ]  **Code Paths & Cleanup**
>   Remove unused conditional statements if no existing configuration requires them. Also, remove the unnecessary config parameters.
>   ```python
>   # Instead of:
>   if config.use_last_layer_norm:
>       self.norm = nn.LayerNorm()
>   else:
>       self.norm = nn.Identity()
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   ```python
>   # If `use_last_layer_norm = True` in all configs, simplify it to:  
>   self.norm = nn.LayerNorm()
>   ```
> * [ ]  **Code Style**
>   
>   * Use clear, descriptive variable names: avoid single-letter (`x`) or placeholder (`tmp`) variables.
>   * Add comments for non-obvious code.
>   * Use type hints for modules (e.g. in `__init__` and `forward`).
> * [ ]  **Conversion Script Standard**
>   Please follow the `mllama` model format for the conversion script. Define a single key mapping using regex (see `rt_detr_v2`, `ijepa`, `mllama`, `superglue` models for reference). This is our new standard for all newly added models.
> 
> Thanks again! Looking forward to the updates 🚀

VladOS95-cyber committed:
small refactoring in tests

VladOS95-cyber committed:
small refactoring in tests

VladOS95-cyber committed:
small refactoring in tests

VladOS95-cyber committed:
fix failing tests, some refactoring

VladOS95-cyber committed:
fix failing tests, some refactoring

VladOS95-cyber committed:
fix failing tests, some refactoring

VladOS95-cyber commented:
Hi @qubvel! I fixed all failing tests except repository consistency, because during resnet modular conversion, the script add additional resnet postfix into configuration import path in modeling_d_fine_resnet file like **from .configuration_d_fine_resnet_resnet import DFineResNetConfig**. I manually fixed that, but honestly, I have no idea why modular conversion script behave like that..

VladOS95-cyber committed:
format

VladOS95-cyber committed:
format

VladOS95-cyber committed:
format

qubvel commented:
Hey @VladOS95-cyber, thanks a lot for the update! I have a sprint, but I will try to review it as soon as have any bandwidth!
Regarding modular conversion, it might have some issues, it's still in active development.

cc @Cyrilvallez re `modular` issues

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+    pass
+
+
+def box_xyxy_to_cxcywh(x: torch.Tensor) -> torch.Tensor:
+    x0, y0, x1, y1 = x.unbind(-1)
+    b = [(x0 + x1) / 2, (y0 + y1) / 2, (x1 - x0), (y1 - y0)]
+    return torch.stack(b, dim=-1)
```
We have `corners_to_center_format` function for that, please reuse existing one

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+    """
+    Decodes edge-distances into bounding box coordinates.
+
+    Args:
+        points (Tensor): (B, N, 4) or (N, 4) format, representing [x, y, w, h],
+                         where (x, y) is the center and (w, h) are width and height.
+        distance (Tensor): (B, N, 4) or (N, 4), representing distances from the
+                           point to the left, top, right, and bottom boundaries.
+
+        reg_scale (float): Controls the curvature of the Weighting Function.
+
+    Returns:
+        Tensor: Bounding boxes in (N, 4) or (B, N, 4) format [cx, cy, w, h].
```
Please update the docstrings everywhere to follow the library format. Additionally, use descriptive names whenever possible to eliminate any ambiguity.

```suggestion
    Args:
        points (`torch.Tensor`):
            (batch_size, num_boxes, 4) or (num_boxes, 4) format, representing [x_center, y_center, width, height]
        distance (`torch.Tensor`): 
            (batch_size, num_boxes, 4) or (num_boxes, 4), representing distances from the point to the left, top, right, and bottom boundaries.
        reg_scale (`float`):
            Controls the curvature of the Weighting Function.

    Returns:
        `torch.Tensor`: Bounding boxes in (batch_size, num_boxes, 4) or (num_boxes, 4) format, representing [x_center, y_center, width, height]
```

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+        init.constant_(self.reg_conf.layers[-1].weight, 0)
+
+    def forward(self, scores, pred_corners):
+        B, L, _ = pred_corners.size()
```
Please no one-letter vars, B-> batch_size, L->length

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+        self.k = k
+        self.reg_max = reg_max
+        self.reg_conf = DFineMLP(4 * (k + 1), hidden_dim, 1, num_layers)
+        init.constant_(self.reg_conf.layers[-1].bias, 0)
+        init.constant_(self.reg_conf.layers[-1].weight, 0)
```
No init in modules, all in _init_weights

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+
+
+class DFineLQE(nn.Module):
+    def __init__(self, k, hidden_dim, num_layers, reg_max):
```
should be initialized from `config` (and other modules as well)

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+        hidden_channels = int(out_channels * expansion)
+        self.conv1 = DFineConvNormLayer(config, in_channels, hidden_channels, 1, 1, activation=activation)
+        self.conv2 = DFineConvNormLayer(config, in_channels, hidden_channels, 1, 1, activation=activation)
+        self.bottlenecks = nn.Sequential(
+            *[DFineRepVggBlock(config, hidden_channels, hidden_channels) for _ in range(num_blocks)]
+        )
```
please avoid Sequential, ModuleList has better Tochcript compatibility in case we will need to override or modify smth in subsequent models

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+        else:
+            self.conv3 = nn.Identity()
+
+    def forward(self, hidden_state):
+        device = hidden_state.device
+        hidden_state_1 = self.conv1(hidden_state)
+        hidden_state_1 = self.bottlenecks(hidden_state_1).to(device)
+        hidden_state_2 = self.conv2(hidden_state).to(device)
+        return self.conv3(hidden_state_1 + hidden_state_2)
```
typehints + no operation in return
```python
...
hidden_state = self.conv3(hidden_state_1 + hidden_state_2)
return hidden_state
```

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+        c2 = config.encoder_hidden_dim
+        c3 = config.encoder_hidden_dim * 2
+        c4 = round(config.hidden_expansion * config.encoder_hidden_dim // 2)
+        self.c = c3 // 2
+        self.cv1 = DFineConvNormLayer(config, c1, c3, 1, 1, activation=act)
+        self.cv2 = nn.Sequential(
+            DFineCSPRepLayer(config, c3 // 2, c4, num_blocks=numb_blocks),
+            DFineConvNormLayer(config, c4, c4, 3, 1, activation=act),
+        )
+        self.cv3 = nn.Sequential(
+            DFineCSPRepLayer(config, c4, c4, num_blocks=numb_blocks),
+            DFineConvNormLayer(config, c4, c4, 3, 1, activation=act),
+        )
+        self.cv4 = DFineConvNormLayer(config, c3 + (2 * c4), c2, 1, 1, activation=act)
```
1) conv1, conv2, conv3, conv4
2) no c1/c2/c3 names

Lets have all of these as separate modules, no need to add ModuleList or Sequential just for two modules

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+        )
+        self.cv4 = DFineConvNormLayer(config, c3 + (2 * c4), c2, 1, 1, activation=act)
+
+    def forward_chunk(self, x):
+        y = list(self.cv1(x).chunk(2, 1))
+        y.extend((m(y[-1])) for m in [self.cv2, self.cv3])
+        return self.cv4(torch.cat(y, 1))
+
+    def forward(self, x):
+        y = list(self.cv1(x).split((self.c, self.c), 1))
+        y.extend(m(y[-1]) for m in [self.cv2, self.cv3])
+        return self.cv4(torch.cat(y, 1))
```
do we need both forward functions? please rewrite to make it clear

smth like this

```
hidden_state = self.conv1(hidden_state)
hidden_state_1, hidden_state_2 = torch.split((self.split_channels, self.split_channels), 1)
...

out_features = [hidden_state_1, hidden_state_2, hidden_state_3, hidden_state_4 ..]
return out_features
```
Longer but cleaner and easier to debug.

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+        return self.cv4(torch.cat(y, 1))
+
+
+class DFineSCDown(nn.Module):
+    def __init__(self, config: DFineConfig, c1, c2, k, s):
+        super().__init__()
+        self.cv1 = DFineConvNormLayer(config, c1, c2, 1, 1)
+        self.cv2 = DFineConvNormLayer(config, c2, c2, k, s, c2)
+
+    def forward(self, x):
+        return self.cv2(self.cv1(x))
```
Same here, please no short names

qubvel reviewed (commented):
Hey @VladOS95-cyber, I made an initial review for `modular_d_fine`. Looks much better, thanks for the update! See the comment below, and all these also related to backbone file as well

VladOS95-cyber commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+            backbone_kwargs=backbone_kwargs,
+        )
+
+        self.backbone_config = backbone_config
+        self.backbone = backbone
+        self.use_pretrained_backbone = use_pretrained_backbone
+        self.use_timm_backbone = use_timm_backbone
+        self.freeze_backbone_batch_norms = freeze_backbone_batch_norms
+        self.backbone_kwargs = backbone_kwargs
+        # encoder
+        self.encoder_hidden_dim = encoder_hidden_dim
+        self.encoder_in_channels = encoder_in_channels
+        self.feat_strides = feat_strides
+        self.encoder_ffn_dim = encoder_ffn_dim
+        self.dropout = dropout
+        self.activation_dropout = activation_dropout
+        self.encode_proj_layers = encode_proj_layers
```
I am not sure but it seems that conversion script does not work properly for config. Why then RTDetrV2Config is implemented almost from the scratch but it has almost all duplicated args/kwargs with RTDetrConfig?

VladOS95-cyber replied:
I checked other modulars, they do the same for configs, inherit from PretrainedConfig and even if there are duplicated things, they are implemented from the scratch

VladOS95-cyber replied:
Ahh ok, got it=) I thouth there migh be a reason for that, actually, in earlier commits I reused all args from RTdetr config but then changed that as saw other implementation, to be align so to speak

VladOS95-cyber commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+        self.reg_scale = reg_scale
+        self.depth_mult = depth_mult
+
+    @classmethod
+    def from_backbone_configs(cls, backbone_config: PretrainedConfig, **kwargs):
+        """Instantiate a [`DFineConfig`] (or a derived class) from a pre-trained backbone model configuration and DETR model
+        configuration.
+
+            Args:
+                backbone_config ([`PretrainedConfig`]):
+                    The backbone configuration.
+
+            Returns:
+                [`DFineConfig`]: An instance of a configuration object
+        """
+        return cls(
+            backbone_config=backbone_config,
+            **kwargs,
+        )
```
Modular does not handle renaming for docstrings, at least for config, or I did smth wrong

qubvel replied:
Ok, isn't it identical to rt-detr method except  RTDetrConfig -> DFineConfig ?

VladOS95-cyber replied:
Methods are definetely the same, yes, and I am going to remove duplication. I think I will even use RtDetrV2, as it is better suitable for modular

VladOS95-cyber commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+        )
+
+
+def deformable_attention_core_func_v2(
+    value: torch.Tensor,
+    value_spatial_shapes,
+    sampling_locations: torch.Tensor,
+    attention_weights: torch.Tensor,
+    num_points_list: List[int],
+    method="default",
+):
+    """
+    Args:
+        value (Tensor): [bs, value_length, n_head, c]
```
agree

VladOS95-cyber commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+    return output.permute(0, 2, 1)
+
+
+class DFineMultiscaleDeformableAttention(nn.Module):
+    def __init__(
+        self,
+        config: DFineConfig,
+        method="default",
+    ):
+        """
+        D-Fine version of multiscale deformable attention
+        """
+        super(DFineMultiscaleDeformableAttention, self).__init__()
```
I am not sure whether it will be convinient to reuse RT-DETRv2 in this case because: 1. constructor from RT-DETRv2 takes DETRv2 config as argument while I reuse RTdetr as it is more convinient for modular. 2. DFineMultiscaleDeformableAttention does not have weights for value_proj and output_proj, so I have to rebuild forward method from the scratch anyway. So, honestly in this case I peref to leave it as is, just remove __reset_parameters and that's it

qubvel replied:
Ok, thanks for clarification 🤗

VladOS95-cyber commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+        return output, attention_weights
+
+
+class DFineGate(nn.Module):
+    def __init__(self, d_model):
+        super(DFineGate, self).__init__()
+        self.gate = nn.Linear(2 * d_model, 2 * d_model)
+        bias = self._bias_init_with_prob(0.5)
+        init.constant_(self.gate.bias, bias)
+        init.constant_(self.gate.weight, 0)
+        self.norm = nn.LayerNorm(d_model)
+
+    def forward(self, x1, x2):
+        gate_input = torch.cat([x1, x2], dim=-1)
+        gates = torch.sigmoid(self.gate(gate_input))
+        gate1, gate2 = gates.chunk(2, dim=-1)
+        return self.norm(gate1 * x1 + gate2 * x2)
+
+    def _bias_init_with_prob(self, prior_prob=0.01):
+        """initialize conv/fc bias value according to a given probability value."""
+        bias_init = float(-math.log((1 - prior_prob) / prior_prob))
+        return bias_init
```
agree

VladOS95-cyber commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+        num_backbone_outs = len(config.decoder_in_channels)
+        decoder_input_proj = []
+        in_channels = config.decoder_in_channels[-1]
+        for _ in range(num_backbone_outs):
+            if config.hidden_size == config.decoder_in_channels[-1]:
+                decoder_input_proj.append(nn.Identity())
+            else:
+                decoder_input_proj.append(
+                    nn.Sequential(
+                        nn.Conv2d(in_channels, config.d_model, kernel_size=1, bias=False),
+                        nn.BatchNorm2d(config.d_model, config.batch_norm_eps),
+                    )
+                )
+        for _ in range(config.num_feature_levels - num_backbone_outs):
+            if config.hidden_size == config.decoder_in_channels[-1]:
+                decoder_input_proj.append(nn.Identity())
+            else:
+                decoder_input_proj.append(
+                    nn.Sequential(
+                        nn.Conv2d(in_channels, config.d_model, kernel_size=3, stride=2, padding=1, bias=False),
+                        nn.BatchNorm2d(config.d_model, config.batch_norm_eps),
+                    )
+                )
```
The problem here is that I completely reuse logic from RTDetrModel including forward function where it is used as Sequential. So, changing decoder_input_proj to ModuleList gives us a lot of complications not only in weights conversion but in forward propagation, as because of this change I have to rebuild the while forward function which I prefer to avoid. What do you think?

qubvel replied:
Ok, let's leave it this way for consistency

VladOS95-cyber commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1231 @@
+        self.decoder = DFineDecoder(config)
+
+
+class DFineForObjectDetection(RTDetrForObjectDetection, DFinePreTrainedModel):
```
Hmm, I think it is difficult to avoid as from RTDetrForObjectDetection we completely reuse forward prop logic while from DFinePreTrainedModel we use custom _init_weights, the same way it is implemeted for RTDetrV2ForObjectDetection and believe for the same reason=)

VladOS95-cyber committed:
refactoring and cleanup after review

VladOS95-cyber committed:
refactoring and cleanup after review

VladOS95-cyber committed:
refactoring and cleanup after review

VladOS95-cyber commented:
Hi @qubvel! I resolved your comments and did some refactoring and cleanup

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,846 @@
+            if config.hidden_size == config.decoder_in_channels[-1]:
+                decoder_input_proj.append(nn.Identity())
+            else:
+                decoder_input_proj.append(
+                    nn.Sequential(
+                        nn.Conv2d(in_channels, config.d_model, kernel_size=1, bias=False),
+                        nn.BatchNorm2d(config.d_model, config.batch_norm_eps),
+                    )
+                )
```
Keeping the same logic, but a bit easier to read IMO
```suggestion
                conv = nn.Conv2d(in_channels, config.d_model, kernel_size=1, bias=False)
                batchnorm = nn.BatchNorm2d(config.d_model, config.batch_norm_eps)
                decoder_input_proj.append(nn.Sequential(conv, batchnorm))
```

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,846 @@
+        layer_scale (`float`, *optional*, defaults to 1.0):
+            Scaling factor for the hidden dimension in later decoder layers. Used to adjust the
+            model capacity after the evaluation layer.
+        reg_max (`int`, *optional*, defaults to 32):
+            Maximum number of bins for the distribution-guided bounding box refinement.
+            Higher values allow for more fine-grained localization but increase computation.
+        reg_scale (`float`, *optional*, defaults to 4.0):
+            Scale factor for the regression distribution. Controls the range and granularity
+            of the bounding box refinement process.
```
Just a suggestion: may be use better name for `reg_max`? smth like `max_num_bins`?
But it's up to you, not feel strong re this

VladOS95-cyber replied:
agree

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,846 @@
+
+class DFineLQE(nn.Module):
+    def __init__(self, config: DFineConfig):
+        super(DFineLQE, self).__init__()
```
```suggestion
        super().__init__()
```

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,846 @@
+        self.reg_max = config.reg_max
+        self.reg_conf = DFineMLP(4 * (self.top_prob_values + 1), config.lqe_hidden_dim, 1, config.lqe_layers)
+
+    def forward(self, scores, pred_corners):
```
Type hints and docstring would be appreciated

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,846 @@
+        )
+
+
+class DFineRepVggBlock(nn.Module):
+    def __init__(self, config: DFineConfig, in_channels: int, out_channels: int):
+        super().__init__()
+
+        activation = config.activation_function
+        self.conv1 = DFineConvNormLayer(config, in_channels, out_channels, 3, 1, padding=1)
+        self.conv2 = DFineConvNormLayer(config, in_channels, out_channels, 1, 1, padding=0)
+        self.activation = nn.Identity() if activation is None else ACT2CLS[activation]()
+
+    def forward(self, hidden_state: torch.Tensor) -> torch.Tensor:
+        hidden_state = self.conv1(hidden_state) + self.conv2(hidden_state)
+        hidden_state = self.activation(hidden_state)
+        return hidden_state
+
```
looks super clear, but seems identical to RT-DETR one? Can we avoid overwriting it here?

```
class DFineRepVggBlock(RTDetrRepVggBlock):
    pass
```

VladOS95-cyber replied:
We can avoid partially yes, but not completely as conv1 and conv2 layers should be D-Fine one DFineConvNormLayer as weights sizes comming from original checkpoint is always different and complicated and we have to control in_channels, out_channels and so on differently depending on model params

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,846 @@
+    Cross Stage Partial (CSP) network layer with RepVGG blocks.
+    """
+
+    def __init__(self, config: DFineConfig, in_channels: int, out_channels: int, num_blocks: int, expansion=1.0):
```
```suggestion
    def __init__(self, config: DFineConfig, in_channels: int, out_channels: int, num_blocks: int, expansion: float = 1.0):
```

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,846 @@
+        # top-down fpn
+        self.lateral_convs = nn.ModuleList()
+        self.fpn_blocks = nn.ModuleList()
+        for _ in range(len(self.in_channels) - 1, 0, -1):
+            self.lateral_convs.append(
+                DFineConvNormLayer(config, self.encoder_hidden_dim, self.encoder_hidden_dim, 1, 1)
+            )
+            self.fpn_blocks.append(DFineRepNCSPELAN4(config, numb_blocks=round(3 * config.depth_mult)))
```
```suggestion
        for _ in range(len(self.in_channels) - 1, 0, -1):
            lateral_layer = DFineConvNormLayer(config, self.encoder_hidden_dim, self.encoder_hidden_dim, 1, 1)
            self.lateral_convs.append(lateral_layer)
            num_blocks = round(3 * config.depth_mult)
            fpn_layer = DFineRepNCSPELAN4(config, numb_blocks=num_blocks)
            self.fpn_blocks.append(fpn_layer)
```

qubvel commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,846 @@
+        self.downsample_convs = nn.ModuleList()
+        self.pan_blocks = nn.ModuleList()
+        for _ in range(len(self.in_channels) - 1):
+            self.downsample_convs.append(
+                nn.Sequential(
+                    DFineSCDown(config, 3, 2),
+                )
+            )
+            self.pan_blocks.append(DFineRepNCSPELAN4(config, numb_blocks=round(3 * config.depth_mult)))
```
Looks weird actually with one layer in a row, can we use just

```suggestion
            self.downsample_convs.append(DFineSCDown(config, 3, 2))
            num_blocks = round(3 * config.depth_mult)
            self.pan_blocks.append(DFineRepNCSPELAN4(config, numb_blocks=num_blocks))
```

VladOS95-cyber replied:
agree

qubvel commented on src/transformers/models/d_fine/modular_d_fine_resnet.py:
```diff
@@ -0,0 +1,414 @@
+from ..rt_detr.modeling_rt_detr_resnet import RTDetrResNetBackbone, RTDetrResNetConvLayer, RTDetrResNetPreTrainedModel
+
+
+class DFineResNetStageConfig(NamedTuple):
+    """
+    Configuration for each stage in the D-FINE ResNet backbone.
+    Each stage is defined by a list of parameters in the following order:
+    [in_channels, mid_channels, out_channels, num_blocks, downsample, light_block, kernel_size, layer_num]
+
+    Args:
+        stage1 (`List[Any]`, defaults to [48, 48, 128, 1, False, False, 3, 6]):
+            First stage configuration:
+            - Input channels: 48
+            - Middle (bottleneck) channels: 48
+            - Output channels: 128
+            - Number of blocks: 1
+            - No downsampling
+            - Standard (non-light) blocks
```
Let's put hit into config directly, no need for separate class

qubvel commented on src/transformers/models/d_fine/modular_d_fine_resnet.py:
```diff
@@ -0,0 +1,414 @@
+        stride: int = 1,
+        groups: int = 1,
+        activation: str = "relu",
+        use_lab: bool = False,
```
use_lab -> use_learnable_affine_block

qubvel commented on src/transformers/models/d_fine/modular_d_fine_resnet.py:
```diff
@@ -0,0 +1,414 @@
+
+
+class DFineResNetEseModule(nn.Module):
+    def __init__(self, chs):
+        super().__init__()
+        self.conv = nn.Conv2d(
+            chs,
+            chs,
+            kernel_size=1,
+            stride=1,
+            padding=0,
+        )
+        self.sigmoid = nn.Sigmoid()
```
```suggestion
    def __init__(self, channels: int):
        super().__init__()
        self.conv = nn.Conv2d(
            channels,
            channels,
            kernel_size=1,
            stride=1,
            padding=0,
        )
        self.activation = nn.Sigmoid()
```

qubvel commented on src/transformers/models/d_fine/modular_d_fine_resnet.py:
```diff
@@ -0,0 +1,414 @@
+                "Make sure that the channel dimension of the pixel values match with the one set in the configuration."
+            )
+        embedding = self.stem1(pixel_values)
+        embedding = F.pad(embedding, (0, 1, 0, 1))
+        emb_stem_2a = self.stem2a(embedding)
+        emb_stem_2a = F.pad(emb_stem_2a, (0, 1, 0, 1))
```
Can we instead use padding in required subsequent conv layers?

qubvel commented on src/transformers/models/d_fine/modular_d_fine_resnet.py:
```diff
@@ -0,0 +1,414 @@
+
+
+class DFineResNetBasicLayer(nn.Module):
+    def __init__(
+        self,
+        in_chs,
+        mid_chs,
+        out_chs,
+        layer_num,
+        kernel_size=3,
+        residual=False,
+        light_block=False,
+        agg="ese",
+        drop_path=0.0,
+        use_lab=False,
+    ):
```
can we initialize form config? and please use more descriptive var names
```
in_chs -> in_channels
...
```

VladOS95-cyber replied:
it is not really convinient to initialize it from config as every stage has its own args which is easier to use iteration over every stage config layer args and initialize object, i don't see ay possible way how to make it simpler as layer param controlling is complex enough. I prefer to leave it as is like here: `for _, stage in enumerate(config.stage_config):
            in_channels, mid_channels, out_channels, block_num, downsample, light_block, kernel_size, layer_num = stage
            self.stages.append(
                DFineResNetStage(
                    in_channels,
                    mid_channels,
                    out_channels,
                    block_num,
                    layer_num,
                    downsample,
                    light_block,
                    kernel_size,
                    use_learnable_affine_block=config.use_learnable_affine_block,
                )
            )`

qubvel replied:
Ok, got it. Please update param names 👍

VladOS95-cyber replied:
sure, it is done!

qubvel commented on src/transformers/models/d_fine/modular_d_fine_resnet.py:
```diff
@@ -0,0 +1,414 @@
+
+        self.layers = nn.ModuleList()
+        for i in range(layer_num):
+            if light_block:
+                self.layers.append(
+                    DFineResNetConvLayerLight(
+                        in_chs if i == 0 else mid_chs, mid_chs, kernel_size=kernel_size, use_lab=use_lab
+                    )
+                )
+            else:
+                self.layers.append(
+                    DFineResNetConvLayer(
+                        in_chs if i == 0 else mid_chs, mid_chs, kernel_size=kernel_size, stride=1, use_lab=use_lab
+                    )
+                )
+
```
```suggestion
            in_channels = in_channels if i == 0 else middle_channels
            if light_block:
                block = DFineResNetConvLayerLight(
                    in_channels=in_channels, 
                    middle_channels=middle_channels,
                    kernel_size=kernel_size,
                    use_learnable_affine_block=use_learnable_affine_block,
                )
            else:
                block = DFineResNetConvLayer(
                    in_channels=in_channels, 
                    middle_channels=middle_channels,
                    kernel_size=kernel_size,
                    use_learnable_affine_block=use_learnable_affine_block,
                    stride=1,
                )
            self.layers.append(block)
```

qubvel commented on src/transformers/models/d_fine/modular_d_fine_resnet.py:
```diff
@@ -0,0 +1,414 @@
+
+        # feature aggregation
+        total_chs = in_chs + layer_num * mid_chs
+        if agg == "se":
```
```suggestion
        if aggregation == "se":
```

qubvel commented on src/transformers/models/d_fine/modular_d_fine_resnet.py:
```diff
@@ -0,0 +1,414 @@
+
+
+class DFineResNetStage(nn.Module):
+    def __init__(
+        self,
+        in_chs,
+        mid_chs,
+        out_chs,
+        block_num,
+        layer_num,
+        downsample=True,
+        light_block=False,
+        kernel_size=3,
+        agg="se",
+        drop_path=0.0,
+        use_lab=False,
+    ):
```
same here re var naming, config init, type hints

qubvel commented on src/transformers/models/d_fine/modular_d_fine_resnet.py:
```diff
@@ -0,0 +1,414 @@
+                    use_lab=use_lab,
+                )
+            )
+        self.blocks = nn.Sequential(*blocks_list)
```
Let change to module list

qubvel reviewed (commented):
Hi @VladOS95-cyber, thanks a lot for iteration on it! I made another pass and it looks much closer to our standards 🤗 See the comments below

VladOS95-cyber commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,846 @@
+
+        super().__init__(**super_kwargs)
+
+        del self.d_model
+        del self.encoder_attention_heads
+
+        if not hasattr(self, "d_model"):
+            self.d_model = d_model
+
+        if not hasattr(self, "encoder_attention_heads"):
+            self.encoder_attention_heads = encoder_attention_heads
+
```
Otherwise, I have no idea how to resolve this problem

qubvel replied:
I would better modify the test, rather than adjust the modeling code with such a trick.. that's definitely look weird and was overlooked in RT-DETRv2

VladOS95-cyber replied:
Yes, agree, I'll take a look what I can do

VladOS95-cyber replied:
@qubvel I found a problem, I was wrong a bit about tests. The problem is in converted config after modular. These failing tests just check setter of some properties which is correct, but when we convert config using modular, the script place super().__init__(is_encoder_decoder=is_encoder_decoder, **kwargs) call for PretrainedConfig constructor in the very beginning of the __init__ of DFineConfig, and it means that when we try to setup some properties(like hidden size which is unit tests about) using kwargs for example, we setup these properties correctly, yes, but then we override them as all self.args assignment happens after we call PretrainedConfig init. So, we setup, then we override and get wrong values basically

VladOS95-cyber replied:
I presume that the problem could be in modular conversion logic as it places super().__init__ calls too early

VladOS95-cyber replied:
So, the question is, what we can do about that?

VladOS95-cyber commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,846 @@
+            multi_scale_deformable_attention_v2, method=self.decoder_method
+        )
+
+        if self.decoder_method == "discrete":
+            for p in self.sampling_offsets.parameters():
+                p.requires_grad = False
```
Actually not, i think we can get rid of it

VladOS95-cyber commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,846 @@
+                module.bias.data.zero_()
+
+        if isinstance(module, DFineGate):
+            bias = float(-math.log((1 - 0.5) / 0.5))
```
not really=)

VladOS95-cyber committed:
some fixes and refactoring after second review

VladOS95-cyber committed:
some fixes and refactoring after second review

VladOS95-cyber committed:
some fixes and refactoring after second review

VladOS95-cyber commented:
Hi @qubvel! Thank you for your comments. I resolved most of them, for several I provided my comments

VladOS95-cyber committed:
couple of small fixes

VladOS95-cyber committed:
couple of small fixes

VladOS95-cyber committed:
couple of small fixes

VladOS95-cyber committed:
some cleanup

VladOS95-cyber committed:
some cleanup

VladOS95-cyber committed:
some cleanup

VladOS95-cyber committed:
Merge branch 'add-d-fine-model' of https://github.com/VladOS95-cyber/transformers into add-d-fine-model

VladOS95-cyber committed:
remove unused import

VladOS95-cyber committed:
remove unused import

VladOS95-cyber commented:
Hi @qubvel! Sorry for the mess with commits, just unlucky rebase..., I hope it is not an obstacle for review. To sum up, I cleaned up modulars according to your comments as much as possible. On this step we have two unresolved problems: 1. Wrong namespace import for **modeling_d_fine_resnet.py** where conversion script provides **from .configuration_d_fine_resnet_resnet**. 2. DFine config conversion using modular does not work properly because it provides a call **super().__init__(is_encoder_decoder=is_encoder_decoder, **kwargs)** too eraly in the very beginning of __init__ and before all self.args assignments, which is wrong as well.

VladOS95-cyber committed:
add some type hints

VladOS95-cyber committed:
add some type hints

damianoamatruda committed:
Fix XGLM loss computation (PyTorch and TensorFlow) (#35878)

* Fix XGLM loss computation (PyTorch and TensorFlow)

* Update expected output string in XGLM sample test

This updates the expected output string of test_xglm_sample for torch
2.0 to the correct one and removes the one for torch 1.13.1 + cu116
(transformers moved to torch 2.0 with PR #35358).

* Update expected output IDs in XGLM generation test

VladOS95-cyber committed:
Merge branch 'add-d-fine-model' of https://github.com/VladOS95-cyber/transformers into add-d-fine-model
